### PR TITLE
Allow permitting numeric params

### LINKED ISSUE #41625: StrongParameters#permit has inconsistent behavior with numeric-looking keys
### Steps to reproduce

```ruby
params = ActionController::Parameters.new(items: {'1': {x: 2, y: 3}, '2': {x: 3, y: 4}})

params.require(:items).permit('1': [:x, :y], '2': [:x])
#  {"1"=>{"x"=>2, "y"=>3}, "2"=>{"x"=>3}}

params.permit(items: {'1': [:x, :y], '2': [:x]})
#  {"items"=>{"1"=>{}, "2"=>{}}}
```

### Expected behavior

Permit should behave the same when called on a parent hash as it does when called on a child hash.

### Actual behavior

Permit's behavior is inconsistent, and it's very difficult and confusing to permit a nested hash with integer-like keys. More generally, when dealing with JSON parameters, the implicit "array-like" handling behavior should be limited to real, actual arrays—not just objects with keys that look like numbers.

### System configuration
**Rails version**: 6.0.2.2

**Ruby version**: 2.6.3

### PR DESCRIPTION
**Pull request by HParker**
When specifying numeric parameters, strong params lets you permit them all using the same permitted params for each.

For example params like,
```ruby
book: {
        authors_attributes: {
          '0': { name: "William Shakespeare", age_of_death: "52" },
          '1': { name: "Unattributed Assistant" },
          '2': "Not a hash",
          'new_record': { name: "Some name" }
        }
      }
```

can be permitted with,
```
permit book: { authors_attributes: [ :name ] }
```

This returns the name keys for each of the numeric keyed params that have a name field,
```ruby
{ "book" => { "authors_attributes" => { "0" => { "name" => "William Shakespeare" }, "1" => { "name" => "Unattributed Assistant" } } } }
```

This is exactly what you want most of the time. Rarely you might need
to specify different keys for particular numeric attributes. This
allows another strong params syntax for those cases where you can
specify the keys allowed for each individual numerically keys attributes
hash.

After this change using the same params above, you can permit the name and age for only the `0` key and only the name for the `1` key,

```ruby
permit book: { authors_attributes: { '1': [ :name ], '0': [ :name, :age_of_death ] } }
```

This returns exactly the parameters that you specify,

```ruby
{ "book" => { "authors_attributes" => { "0" => { "name" => "William Shakespeare", "age_of_death" => "52" }, "1" => { "name" => "Unattributed Assistant" } } } }
```

Sidenote: this allows `permit` to do the equivalent to
```ruby
params.require(:book).permit(authors_attributes: { '1': [:name]})
```

without raising when `book: ... ` is not present.

The simpler syntax should be preferred, but in cases where you need more control, this is a nice option to have.

Fixes: https://github.com/rails/rails/issues/41625

### CODE REVIEW

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 1]
```diff
@@ -1052,7 +1095,6 @@ def non_scalar?(value)
       EMPTY_HASH  = {}
       def hash_filter(params, filter)
         filter = filter.with_indifferent_access
-
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/strong_parameters.rb:1055]: Please keep this line. It is to visually separate the block

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 2]
```diff
@@ -592,6 +592,43 @@ def require(key)
...
+    #         phone: "555-1234"
+    #       },
+    #       '1': {
+    #         email: "nothing@test.com",
+    #         phone: "555-6789"
+    #       },
+    #     }
+    #   })
+    #   params.permit(person: [:email, :phone])
+    #   # => #<ActionController::Parameters {"person"=>#<ActionController::Parameters {"0"=>#<ActionController::Parameters {"email"=>"none@test.com", "phone"=>"555-1234"} permitted: true>, "1"=>#<ActionController::Parameters {"email"=>"nothing@test.com", "phone"=>"555-6789"} permitted: true>} permitted: true>} permitted: true>
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/strong_parameters.rb:612]: This line is too big. Let's break it.

File: actionpack/lib/action_controller/metal/strong_parameters.rb [hunk 3]
```diff
@@ -592,6 +592,43 @@ def require(key)
...
+    #       },
+    #       '1': {
+    #         email: "nothing@test.com",
+    #         phone: "555-6789"
+    #       },
+    #     }
+    #   })
+    #   params.permit(person: [:email, :phone])
+    #   params.permit(person: { '0': [:email], '1': [:phone]})
+=> #<ActionController::Parameters {"person"=>#<ActionController::Parameters {"0"=>#<ActionController::Parameters {"email"=>"none@test.com"} permitted: true>, "1"=>#<ActionController::Parameters {"phone"=>"555-6789"} permitted: true>} permitted: true>} permitted: true>
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/strong_parameters.rb:630]: This line is invalid and also too big.

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

**Comment by HParker** [comment]
Thanks for the feedback @rafaelfranca and thanks for catching my doc mistakes! ❤️

**Comment by zzak** [comment]
@HParker Do you think this should have had a CHANGELOG?

**Pull request merged**